monai/apps/pathology/datasets.py

```diff
@@ -160,3 +166,132 @@ def __init__(
+            Defaults to CuCIM.
+
+    """
+
+    def __init__(
+        self,
+        data: List,
+        patch_size: Union[int, Tuple[int, int]],
+        transform: Optional[Callable] = None,
+        image_reader_name: str = "cuCIM",
+    ) -> None:
+        super().__init__(data, transform)
+
+        if isinstance(patch_size, int):
+            self.patch_size = np.array((patch_size, patch_size))
+        else:
+            self.patch_size = np.array(patch_size)
+        self.image_reader_name = image_reader_name
+        self.image_reader = WSIReader(image_reader_name)
+
+        # process data and create a list of dictionaries containing all required data and metadata
+        self.data_list = self._create_data_list(data)
+
+        # calculate cummulative number of patches for all whole slide images
+        self.cum_num_patches = np.cumsum([0] + [len(d["image_locations"]) for d in self.data_list])
+        self.total_num_patches = self.cum_num_patches[-1]
+        self.cum_num_patches = self.cum_num_patches[:-1]
+
+    def _create_data_list(self, data: List[Dict]) -> List[Dict]:
+        data_list = []
+        print("Number of whole slide images: ", len(data))
+        for sample in data:
+            processed_data = self._preprocess_sample(sample)
+            data_list.append(processed_data)
+        return data_list
+
+    def _preprocess_sample(self, sample: Dict) -> Dict:
+        """
+        Preprocess input data to load WSIReader object and the foreground mask,
+        and define the locations where patches need to be extracted.
+        """
+        image = self.image_reader.read(sample["image"])
+        mask = np.load(sample["label"]).T
+        try:
+            level, ratio = self._calculate_mask_level(image, mask)
+        except ValueError as err:
+            err.args = (sample["label"],) + err.args
+            raise
+        print(f"Mask ({sample['label']}) at level {int(level)}, with ratio {int(ratio)}")
+
+        # get all indices for tissue region from the foreground mask
+        # note: output same size as the foreground mask and not original wsi image size
+        mask_locations = np.vstack(mask.nonzero()).T
+
+        # convert mask locations to image locations to extract patches
+        image_locations = (mask_locations + 0.5) * ratio - self.patch_size // 2
+
+        return {
+            "name": os.path.splitext(os.path.basename(sample["image"]))[0],
+            "image": image,
+            "mask_shape": mask.shape,
+            "mask_locations": mask_locations.astype(int).tolist(),
+            "image_locations": image_locations.astype(int).tolist(),
+            "level": level,
+        }
+
+    def _calculate_mask_level(self, image, mask) -> Tuple[int, int]:
+        """Calculate level of the mask and its ratio with respect to the whole slide image"""
+        dim_y_img, dim_x_img, _ = image.shape
```

Comment:
Hi @behxyz , I think shape length is needed here, or please add some doc strings to disclose the spatial shapes for input images and masks. What do you think?

Reply:
missing types for this and the following functions.

Reply:
Sure, thanks

Reply:
I addressed it in #1869